### PR TITLE
Mode S surveillance identity

### PR DESCRIPTION
**Pull request by wiseman**
I'm working on adding decoding of more message types. This one handles the Mode S "surveillance identity" response to get the transponder squawk code.

I'm a Rust beginner, and if possible I'd love your feedback on some things. I'm sure there are lots of small issues with my code, and I'd welcome any advice. A larger issue might be how I added `ModeSMessage` as another variant of `MessageKind`; does that seem right?  It feels a little odd to have nested `kind` fields.

Thanks for your time.

### CODE REVIEW

File: src/crc.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+impl fmt::Display for CrcError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Error for CrcError {}
+
+pub fn mode_s_crc(input: &[u8], num_bytes: u8) -> Result<u32, CrcError> {
```
Comment (asmarques) [inline src/crc.rs]: As this is just used internally I would switch this to `pub(crate)`

File: src/lib.rs [hunk 2]
```diff
@@ -6,8 +6,13 @@
 //!   - **TC 19**: Airborne velocity
 
 pub mod cpr;
+mod crc;
 mod parser;
 mod types;
 
+pub use crc::*;
```
Comment (asmarques) [inline src/lib.rs]: See other comment about `pub(crate)`

File: src/parser.rs [hunk 3]
```diff
@@ -198,10 +199,94 @@ fn parse_unknown(input: (&[u8], usize)) -> IResult<(&[u8], usize), MessageKind>
...
+    if f & 0x0200 != 0 {
+        hex_gillham |= 0x2000;
+    } // Bit  9 = A2
+    if f & 0x0100 != 0 {
+        hex_gillham |= 0x0040;
+    } // Bit  8 = C4
+    if f & 0x0080 != 0 {
+        hex_gillham |= 0x4000;
+    } // Bit  7 = A4
+      //if (ID13Field & 0x0040) {hexGillham |= 0x0800;} // Bit  6 = X  or M
```
Comment (asmarques) [inline src/parser.rs:234]: Any reason for not handling this case?

File: src/types.rs [hunk 4]
```diff
@@ -89,10 +114,21 @@ pub enum MessageKind {
...
+    },
     /// Unsupported message
     Unknown,
 }
 
+#[derive(Debug, PartialEq, Clone)]
+pub enum ModeSMessageKind {
+    // DF=5
+    SurveillanceIdentity { squawk: Squawk },
+    Unknown,
```
Comment (asmarques) [inline src/types.rs]: Parsing of unsupported messages, both ADSB messages and these new Mode-S messages, is already caught by `MessageKind::Unknown` so this case isn't required.

File: src/parser.rs [hunk 5]
```diff
@@ -198,10 +199,94 @@ fn parse_unknown(input: (&[u8], usize)) -> IResult<(&[u8], usize), MessageKind>
...
+    ))
+}
+
+fn parse_mode_s_message_kind(input: (&[u8], usize)) -> IResult<(&[u8], usize), ModeSMessageKind> {
+    parse_surveillance_identity(input)
+}
+
+fn parse_mode_s_message(input: (&[u8], usize)) -> IResult<(&[u8], usize), MessageKind> {
+    let (input, kind) = parse_mode_s_message_kind(input)?;
+    let crc = mode_s_crc(input.0, 7).unwrap();
```
Comment (asmarques) [inline src/parser.rs]: This shouldn't panic. Unfortunately I haven't found a good way to handle custom parsing errors using nom so something like `.map_err(|| Err::Error(make_error(input, ErrorKind::LengthValue)))?;`, while not ideal,  would at least avoid crashing here.

File: src/crc.rs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    let mut rem = 0u32;
+    if num_bytes < 3 {
+        return Err(CrcError(String::from(
```
Comment (asmarques) [inline src/crc.rs:41]: Please see my other comment where this is called. Given the lack of custom error handling in the parsers I would replace the `CrcError` type with a simple `Option` for now. I want to improve the surfacing of custom parsing errors but that shouldn't block merging this.

**Comment by asmarques** [comment]
Great to see more message types supported. Since all fields after the DF are different, adding a new `MessageKind::ModeSMessage` and `ModeSMessageKind` seems like the best way to handle it.

File: src/parser.rs (hunk 3 above)
Comment (wiseman) [reply src/parser.rs:234]: That bit appears to be a "don't care" bit.  From https://mode-s.org/decode/content/mode-s/3-surveillance.html:
 
> The 13-bit identity code encodes the 4 octal digit squawk code (from 0000 to 7777). The structure of this field is shown as follows:
> 
> ```
> +----+----+----+----+----+----+---+----+----+----+----+----+----+
> | C1 | A1 | C2 | A2 | C4 | A4 | X | B1 | D1 | B2 | D2 | B4 | D4 |
> +----+----+----+----+----+----+---+----+----+----+----+----+----+
> ```
> 
> The binary representation of the octal digit is:
> ```
> A4 A2 A1 | B4 B2 B1 | C4 C2 C1 | D4 D2 D1
> ```
>

File: src/lib.rs (hunk 2 above)
Comment (wiseman) [reply src/lib.rs]: Done.

**Pull request merged**